rest_framework/test.py

```diff
@@ -385,14 +393,20 @@ def setUpClass(cls):
         cls._module.urlpatterns = cls.urlpatterns
 
         cls._override.enable()
+
+        if django.VERSION > (4, 0):
+            cls.addClassCleanup(cls._override.disable)
+            cls.addClassCleanup(cleanup_url_patterns, cls)
```

Comment:
Is `addClassCleanup` a new feature in Django 4.0?

I'm not seeing any differences in the docs re. the existing `tearDownClass`... https://docs.djangoproject.com/en/4.0/topics/testing/tools/#simpletestcase

Reply:
`addClassCleanup` is from `unittest.TestCase`, added in Python 3.8 https://docs.python.org/3.9/library/unittest.html#unittest.TestCase.addClassCleanup

Reply:
Django just happens to be using it from Django 4.0 — See https://github.com/django/django/commit/faba5b702a9c5bb9452a543100928bcb5f66ebcf

Reply:
Ah fab. How about we switch based on the Python version, rather than the Django version?

Reply:
Maybe we should comment on the release notes for this? 🤔 (cc @felixxm )

Reply:
> Maybe we should comment on the release notes for this? thinking (cc @felixxm )

Sure, if you think it might be helpful. Is it not enough to call `super().setUpClass()` at the beginning? :thinking: 
```python
    @classmethod
    def setUpClass(cls):
        super().setUpClass()
        # Get the module of the TestCase subclass
        cls._module = import_module(cls.__module__)
        ...
```

Reply:
It might be... Let me try that after lunch 🥪

Reply:
... or use `cls._overridden_settings` and remove redundant `cls._override.enable()/disable()` calls :thinking: 
```python
cls._overridden_settings = {
    'ROOT_URLCONF': cls.__module__,
    **cls._overridden_settings,
}
super().setUpClass()
```

Reply:
Moving `super().setUpClass()` from the *bottom* of the `setUpClass` method to the top resolves this without any `4.0` branching.

Do I properly understand why? Nope.

Reply:
In other words, this works...

```python
    @classmethod
    def setUpClass(cls):
        super().setUpClass()  # <--- This line has moved.

        # Get the module of the TestCase subclass
        cls._module = import_module(cls.__module__)
        cls._override = override_settings(ROOT_URLCONF=cls.__module__)

        if hasattr(cls._module, 'urlpatterns'):
            cls._module_urlpatterns = cls._module.urlpatterns

        cls._module.urlpatterns = cls.urlpatterns

        cls._override.enable()

    @classmethod
    def tearDownClass(cls):
        super().tearDownClass()
        cls._override.disable()

        if hasattr(cls, '_module_urlpatterns'):
            cls._module.urlpatterns = cls._module_urlpatterns
        else:
            del cls._module.urlpatterns
```